### PR TITLE
Improving performance

### LINKED ISSUE #80: Weeder runs sequentially
Weeder does not take advantage of concurrent Haskell.

At Freckle we run our builds on multi core machines. If a build is going slower than we need, increasing the size of the machine is a relatively cheap way to gain performance.

`weeder` is a fly in that ointment. It currently takes 4 minutes to complete on our primary repository and we cannot "throw money" at the problem by increasing concurrency.

Introducing concurrency into `weeder` would allow it to gain performance in a typical GHC build scenario (multi core machines).

### PR DESCRIPTION
**Pull request by ryndubei**
Applied on top of #134. Resolves #80.

This PR includes several performance optimisations: see the commit descriptions for information about individual ones.

The important commits here are the following:

- 77a2ed69b2cd6ba7f6b046852d2442171cc24ffd
- 3259b4643268426e9f3279b7232ef6c001d092d6
- bb4edbc7f8a141acc3b430a788c1c35ce3cd62f6
- a4ae9bd6c50f76eb052f718fb4a8025b809494ce
- 5f7a8d9296190d2c2899b5c2556d2ce22cf7d2b7 with 7655729b3d0e7aecbcffac15e4e13849d48f14a9

The most significant is a4ae9bd6c50f76eb052f718fb4a8025b809494ce, which adds parallelism. This can almost halve execution time when given enough cores.

The commits before a4ae9bd6c50f76eb052f718fb4a8025b809494ce are mostly just enough to cancel out the overhead of #126: the execution time of 2.6.0 is close to this PR when given only one core with identical settings.

## Benchmarks

Using HIE files generated from compiling GHC. Default values for `roots` and `root-instances`.

#### Current figures of #134

`type-class-roots = true`, `unused-types = false`:

```
real    0m14.458s
user    0m13.478s
sys     0m0.953s
```

`type-class-roots = false`, `unused-types = true`:

```
real    0m24.698s
user    0m21.836s
sys     0m2.831s
```

#### This PR, -N1

`type-class-roots = true`, `unused-types = false`:

```
real    0m9.707s
user    0m8.745s
sys     0m0.931s
```

`type-class-roots = false`, `unused-types = true`:

```
real    0m13.719s
user    0m12.822s
sys     0m0.845s
```

#### This PR, -N12

`type-class-roots = true`, `unused-types = false`:

```
real    0m5.168s
user    0m14.584s
sys     0m1.752s
```

`type-class-roots = false`, `unused-types = true`

```
real    0m8.649s
user    0m19.747s
sys     0m1.857s
```

### CODE REVIEW

**Comment by ocharles** [comment]
Wow, this is absolutely massive - 4 minutes at work drops down to just over 1 minute, with `-j8`!

File: src/Weeder.hs [hunk 1]
```diff
@@ -158,7 +162,7 @@ data Analysis =
   Analysis
     { dependencyGraph :: Graph Declaration
       -- ^ A graph between declarations, capturing dependencies.
-    , declarationSites :: Map Declaration ( Set RealSrcSpan )
+    , declarationSites :: Map Declaration (Set Int)
```
[the remark this answers is not in the document]
Comment (ocharles) [reply src/Weeder.hs:165]: Ah, it's just line number now.

**Review by ocharles — APPROVED** [review]
Great work! Let's get the depend PRs merged in first, but otherwise this is good to go.

File: src/Weeder.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryndubei) [reply src/Weeder.hs:165]: I updated these comments on the `performance-3` branch, but forgot to add the commit here. The comment now refers to it being a line number.

**Pull request merged**